File: model/src/main/java/net/contrapt/auction/model/AbstractEntity.java
package net.contrapt.auction.model;

import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;
import org.apache.commons.lang.builder.ReflectionToStringBuilder;
import org.apache.commons.lang.builder.ToStringStyle;
import org.springframework.data.annotation.CreatedBy;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedBy;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import java.util.Date;

/**
 * Created by msimmons on 10/9/14.
 */
@MappedSuperclass
@EntityListeners(AuditingEntityListener.class)
public abstract class AbstractEntity {

    @Id
    @GeneratedValue
    private Long id;

    @Version
    private int version;

    @CreatedBy
    @Column(name = "created_by")
    private String createdBy;

    @CreatedDate
    @Column(name = "created_at")
    private Date createdAt;

    @LastModifiedBy
    @Column(name = "updated_by")
    private String updatedBy;

    @LastModifiedDate
    @Column(name = "updated_at")
    private Date updatedAt;

    public Long getId() {
        return id;
    }

    public Integer getVersion() {
        return version;
    }

    public String getCreatedBy() {
        return createdBy;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public String getUpdatedBy() {
        return updatedBy;
    }

    public Date getUpdatedAt() {
        return updatedAt;
    }

    /**
     * No arg constructor for serialization etc
     */
    protected AbstractEntity() {}

    /**
     * Each entity should define its unique business key so that this abstract class can implement hash code and
     * equals based on that
     * @return A unique key string for the entity
     */
    public abstract String uniqueKey();

    @Override
    public String toString() {
        return ReflectionToStringBuilder.toString(this, ToStringStyle.SHORT_PREFIX_STYLE);
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(uniqueKey()).toHashCode();
    }

    @Override
    public boolean equals(Object o) {
        if ( o == null || o.getClass() != getClass() ) return false;
        AbstractEntity other = (AbstractEntity)o;
        return new EqualsBuilder()
              .append(uniqueKey(), other.uniqueKey())
              .isEquals();
    }

}

File: model/src/main/kotlin/net/contrapt/auction/model/AbstractEntity.kt
package net.contrapt.auction.model

import org.apache.commons.lang.builder.EqualsBuilder
import org.apache.commons.lang.builder.HashCodeBuilder
import org.springframework.data.annotation.CreatedBy
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedBy
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.util.*
import javax.persistence.*

/**
 * Created by msimmons on 10/9/14.
 */
@MappedSuperclass
@EntityListeners(AuditingEntityListener::class)
abstract class AbstractEntity : AbstractModel() {

    @Id
    @GeneratedValue
    var id: Long? = null

    @Version
    val version: Int = 0

    @CreatedBy
    @Column(name = "created_by")
    lateinit var createdBy: String
        private set

    @CreatedDate
    @Column(name = "created_at")
    lateinit var createdAt: Date
        private set

    @LastModifiedBy
    @Column(name = "updated_by")
    lateinit var updatedBy: String
        private set

    @LastModifiedDate
    @Column(name = "updated_at")
    lateinit var updatedAt: Date
        private set

    /**
     * Each entity should define its unique business key so that this abstract class can implement hash code and
     * equals based on that
     * @return A unique key string for the entity
     */
    abstract fun uniqueKey(): String

    override fun hashCode(): Int {
        return HashCodeBuilder().append(uniqueKey()).toHashCode()
    }

    override fun equals(other: Any?): Boolean {
        if ( other !is AbstractEntity ) return false
        if ( other.javaClass != javaClass ) return false
        return EqualsBuilder().append(uniqueKey(), other.uniqueKey()).isEquals
    }

}
